### PR TITLE
[PRIMAUK-3186]: Inconsistent http client spans instrumentation due to usage of different Telepoison implementations

### PR DESCRIPTION
**Pull request by michael-a-grammar**
* Updates based off this [PR](https://github.com/primait/prima-pulumi/pull/96#issuecomment-1280117786) comment
* Adds the ability to explicitly set the `http.route` Open Telemetry metadata attribute with a sensible default if not set
* Updates `Telepoison.request` documentation

**TODO:**

- [x] Add tests

* Added

- [x] How does `Telepoison.get` etc. interact with `Telepoison.request`, as `Telepoison.request` doesn't seem to be explicitly called often within Stonehenge?

* Via meta magic and function overrides

- [x] Remove the `http.url` Open Telemetry metadata attribute, as this may contain sensitive data?
* Kept for now

https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PRIMAUK-3186

### CODE REVIEW

File: lib/telepoison.ex [hunk 1]
```diff
@@ -35,20 +35,67 @@ defmodule Telepoison do
...
+          |> URI.parse()
+          |> then(fn uri ->
+            uri.path
+            |> String.split("/")
+            # Skips the first item which is an empty string and the last entry which may contain dynamic data
+            # e.g. user id etc.
+            |> Enum.slice(1..-2)
+            |> Enum.join("/")
+            |> (&"#{&1}/:subpath").()
+          end)
```
Comment (LeartS) [inline lib/telepoison.ex:77]: Split into its own function, e.g. `infer_resource_from_path(`, and add some tests for it?

```elixir
assert infer_resource_from_path("/user") == "/user"
assert infer_resource_from_path("/user/12") == "/user/:subpath"
assert infer_resource_from_path("/user/12/details") == "/user/:subpath"
assert infer_resource_from_path("/") == "/"
```

Also, I don't think only the last entry may contain dynamic data. Ultimately it's a tradeoff how "good" the inferred route would be vs the risk of having dynamic params in there. I suggested that just taking the first path component is very unlikely to contain dynamic parameters (e.g. noone uses non-prefixed routes that start directly with a parameter like `/12/details`) but there may be better heuristics.

File: lib/telepoison.ex [hunk 2]
```diff
@@ -35,20 +35,67 @@ defmodule Telepoison do
...
+      ...> body: "{\"foo\": 3}",
+      ...> headers: [{"Accept", "application/json"}]}
+      iex> Telepoison.request(request, resource_route: "users/edit/:user_id")
+      {:ok, response}
+
+  """
+  def request(%Request{options: request_opts} = request, opts \\ []) do
+    span_name = Keyword.get_lazy(request_opts, :ot_span_name, fn -> compute_default_span_name(request) end)
+
+    [env: env, service: service] = Logger.metadata() |> Keyword.take([:service, :env])
```
Comment (LeartS) [inline lib/telepoison.ex:62]: I don't think Telepoison should rely on any specific Logger metadata to having been set!

IMO Telepoison shouldn't try to find and set `service` and `env` at all. The callers can already specify any additional attribute they want to have by passing it to calls to Telepoison.

And for these "global tags" that are always fixed in a deployment, the opentelemetry exporter and opentelemetry collector can also be configured to automatically add some global tags like service and env to all traces.

File: lib/telepoison.ex [hunk 3]
```diff
@@ -35,20 +35,67 @@ defmodule Telepoison do
...
-      ] ++ Keyword.get(opts, :ot_attributes, [])
+        # TODO: Should this be removed as it may contain sensitive data?
+        {"http.url", request.url},
+        {"service.name", service},
+        {"http.route", path},
+        {"service", service},
+        {"env", env}
+      ] ++ Keyword.get(request_opts, :ot_attributes, [])
 
+    save_parent_ctx()
```
Comment (LeartS) [inline lib/telepoison.ex:94]: Why did you move `save_parent_ctx` all the way down here?

File: lib/telepoison.ex [hunk 4]
```diff
@@ -35,20 +35,67 @@ defmodule Telepoison do
...
     new_ctx = Tracer.start_span(span_name, %{kind: :client, attributes: attributes})
     Tracer.set_current_span(new_ctx)
 
-    super(request)
+    super(request, [])
```
Comment (LeartS) [inline lib/telepoison.ex:98]: Can you explain this change?

File: lib/telepoison.ex [hunk 5]
```diff
@@ -35,20 +35,67 @@ defmodule Telepoison do
...
       [
         {"http.method", request.method |> Atom.to_string() |> String.upcase()},
-        {"http.url", request.url}
-      ] ++ Keyword.get(opts, :ot_attributes, [])
+        # TODO: Should this be removed as it may contain sensitive data?
+        {"http.url", request.url},
+        {"service.name", service},
+        {"http.route", path},
+        {"service", service},
+        {"env", env}
```
Comment (LeartS) [inline lib/telepoison.ex:91]: I don't think Telepoison should set these (`env`, `service` and `service.name`) unless specifically passed in `ot_attributes` (see my comment above).

**Review by LeartS — COMMENTED** [review]
A few comments, thanks for getting the work started on this!

File: lib/telepoison.ex (hunk 1 above)
Comment (milquetoastable) [reply lib/telepoison.ex]: - Updated README.md
- Added tests
- Updated the `:resource_route` inference to essentially grab the first slug on the URL (if present! The above scenarios pass are covered by the tests)
- Moved the `:resource_route` inference logic to its own module

File: lib/telepoison.ex (hunk 3 above)
Comment (milquetoastable) [reply lib/telepoison.ex]: Apologies, this was based on a misunderstanding with how `HTTPoison` works

File: lib/telepoison.ex (hunk 4 above)
Comment (milquetoastable) [reply lib/telepoison.ex]: Apologies, this was based on a misunderstanding with how `HTTPoison` works

File: lib/telepoison.ex (hunk 5 above)
Comment (milquetoastable) [reply lib/telepoison.ex]: Removed, I've added a ticket regarding these proposed changes [here](https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PRIMAUK-3313/Telepoison-add-ability-to-set-global-configuratIon-for-standard-OT-attributes)

File: lib/telepoison.ex (hunk 2 above)
Comment (milquetoastable) [reply lib/telepoison.ex]: Removed!

File: lib/uri.ex [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      count when count > 1 ->
+        segments
+        |> Enum.take(1)
+        |> (&"/#{&1}/:subpath").()
+
+      _ ->
+        @default_route
+    end
+  end
+end
```
Comment (LeartS) [inline lib/uri.ex:43]: Is this module and all these function clauses really necessary? It makes it hard to follow and a bit on the "defensive" side IMO.

We call this from the `request` callback, which is called from the basic request implementation https://github.com/edgurgel/httpoison/blob/bdc665199a349981dcda6aeb227587c835597e7a/lib/httpoison/base.ex#L357-L364 with a Request object with a processed url. At that point we'll always have a Request object with an url, not even sure if a `nil` url is possible at that point, but even if it is I would make it a bit clearer with something like this:

```elixir
@spec infer_route_from_request(HTTPoison.Request.t()) :: binary()
def infer_route_from_request(HTTPoison.Request{url: url}) when is_binary(url), do: infer_route_from_url(url)
def infer_route_from_request(HTTPoison.Request{}), do: @default_route

@spec infer_route_from_url(binary()) :: binary()
def infer_route_from_url(url) when is_binary(url) do
    url
    |> URI.parse()
    |> Map.get(:path, "/")
    |> String.split("/", parts: 2, trim: true)
    |> case do
      [] -> @default_route
      [main_path] -> "/" <> main_path
      [main_path, sub_path] -> "/" <> main_path <> "/:subpath"
    end
  end
```

File: test/uri_test.exs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    test "Request URL '/user/edit/24' is inferred as a route of '/user/:subpath'" do
+      request = %Request{url: "/user/edit/24"}
+
+      result = UtilsURI.infer_route(request)
+
+      assert result == "/user/:subpath"
+    end
+
+    test "Request URL '/user/24' is inferred as a route of '/user:subpath'" do
```
Comment (LeartS) [inline test/uri_test.exs:115]:
```suggestion
    test "Request URL '/user/24' is inferred as a route of '/user/:subpath'" do
```

**Review by LeartS — COMMENTED** [review]
Looks good! Just a couple of minor comments.

More high-level, given that we got asked to move telepoison in the official opentelemetry-contrib repository, and that setting `http.route` on client web spans is not "super-compliant" according to the OpenTelemetry semantic attributes conventions, I'm wondering if maybe we should make the inferral of `http.route` opt-in . e.g. with `resource_route: :infer`. If not specified, no such attribute is added. But that can be for a later PR.

**Comment by milquetoastable** [comment]
> Looks good! Just a couple of minor comments.
> 
> More high-level, given that we got asked to move telepoison in the official opentelemetry-contrib repository (#81), and that setting `http.route` on client web spans is not "super-compliant" according to the OpenTelemetry semantic attributes conventions, I'm wondering if maybe we should make the inferral of `http.route` opt-in . e.g. with `resource_route: :infer`. If not specified, no such attribute is added. But that can be for a later PR.
> 
> @tsloughter what do you think?

Hey @LeartS, I've updated the inference to be opt-in rather than magic - wasn't too much effort to do so!

File: test/uri_test.exs (hunk 7 above)
Comment (milquetoastable) [reply test/uri_test.exs]: Updated

File: lib/uri.ex (hunk 6 above)
Comment (milquetoastable) [reply lib/uri.ex]: Certainly, updated

File: lib/telepoison.ex [hunk 8]
```diff
@@ -35,14 +35,47 @@ defmodule Telepoison do
...
+        route ->
+          route
+      end
+
     attributes =
       [
         {"http.method", request.method |> Atom.to_string() |> String.upcase()},
-        {"http.url", request.url}
+        {"http.url", request.url},
+        {"http.route", resource_route}
```
Comment (LeartS) [inline lib/telepoison.ex:78]: This will set `{"http.route", nil}` if not provided. Not sure what OpenTelemetry then does with that, does it automatically drop attributes with `nil` values? In any case, maybe we should not set it at all when `resource_route` is not set.

File: test/telepoison_test.exs [hunk 9]
```diff
@@ -59,6 +59,20 @@ defmodule TelepoisonTest do
     assert {"app.callname", "mariorossi"} in elem(attributes, 4)
   end
 
+  test "resource route can be explicitly passed to Telepoison invocation" do
+    Telepoison.get!("http://localhost:8000/user/edit/24", [], resource_route: "/user/edit")
+
+    assert_receive {:span, span(attributes: attributes)}, 1000
+    assert {"http.route", "/user/edit"} in elem(attributes, 4)
```
Comment (LeartS) [inline test/telepoison_test.exs:66]: We may introduce some more standard attributes in Telepoison in the future, in which case this test may start failing not because the functionality is broken but just because the `http.route` attribute may not be number 4 anymore. Maybe it's worth rewriting it in a way that doesn't rely on the specific position? We only care about it's presence and value.

**Review by LeartS — COMMENTED** [review]
Looking good, almost ready to go IMO!

I would keep the `ot_` prefix for opentelemetry-related keyword args, like we have for the others (`ot_span_name` and `ot_attributes`), so I would name the new one `ot_resource_route`. What do you think?

**Comment by tsloughter** [comment]
Could you make places that you need custom `infer` logic take functions instead and keep the logic separate?

Like in the Tesla library we now take a function that gets the Tesla request struct for the span name, so by default it does the required defaults of the otel semantic conventions, but the user defined replacement can be more than just a static string.

File: test/telepoison_test.exs (hunk 9 above)
Comment (milquetoastable) [reply test/telepoison_test.exs:66]: I've updated some of the tests that rely on this - can look at the others if required, want to check if the update is okay first!

File: lib/telepoison.ex (hunk 8 above)
Comment (milquetoastable) [reply lib/telepoison.ex]: @LeartS certainly - updated the attribute name and how the `http.route` attribute is added in; not sure if it's a bit too ugly however. What do you think?

**Comment by milquetoastable** [comment]
> Could you make places that you need custom `infer` logic take functions instead and keep the logic separate?
> 
> Like in the Tesla library we now take a function that gets the Tesla request struct for the span name, so by default it does the required defaults of the otel semantic conventions, but the user defined replacement can be more than just a static string.

Hi @tsloughter, sorry not too sure I follow - to confirm, should the inference logic be provided by way of a function to `Telepoison.request` (and the functions that wrap it) via the `:ot_resource_route` option?

**Comment by tsloughter** [comment]
I might not understand the discussion (haven't read all of it), it sounded like you needed attributes separate from the semantic conventions and/or custom values for attributes in the semantic conventions. So my suggestion was to accept a function from the user that can return these, removing the need to do it in the library itself.

**Comment by LeartS** [comment]
@tsloughter I think the user-provided infer function could be a really useful feature to supply as a global / per-instance default, but I don't see the value of taking it as a parameter on the single Telepoison request. At that point, most likely the developer already knows what the resource is and would just pass the specific value, but even if they didn't and had their own logic to calculate it, they could just call the function directly and pass the result as the value of the option.

e.g.
```elixir
def infer_resource(url) do
  # Some logic to infer the resource (/ `http.route` attribute) from the url...
end

# Why would a user do this on a single request
resp = Telepoison.get(avatar_url, ot_resource_route: &infer_resource/1)

# Instead of just calling the function directly
resp = Telepoison.get(avatar_url, ot_resource_route: infer_resource(avatar_url))

# Or even pass the value directly when it's known (which it will be a lot of the times)
resp = Telepoison.get(avatar_url, ot_resource_route: "/users/:id/avatar")
```

I guess it could make sense if the inferral function is based on some complex object that is only available within Telepoison (e.g. the fully built `HTTPoison.Request` object), but that seems a bit overkill.

---------------------

On the other hand, allowing users of the library to supply their own resource route inferral logic at the instance level would be useful, saving them from having to specify `ot_resource_route` in every call to Telepoison, e.g.

```elixir
Telepoison.configure(infer_resource: :disabled)  # no route inferral
Telepoison.configure(infer_resource: :default)  # use the conservative telepoison-provided inferral algorithm

def infer_resource(url) do
  # Some custom logic to infer the resource (/ `http.route` attribute) from the url...
end
Telepoison.configure(infer_resource: &infer_resource/1)  # Bring-your-own inferral algorithm
```

But maybe that could be part of a different PR?

Anyway being elixir it's very easy to have polymorphic behaviour on the param:
* If a string, use that string
* If `:infer`, use the configured inferral function or Telepoison-provided fallback if no custom function has been configured.
* If a function, call the function and use the output

In any case, I'd like to keep a conservative/safe default inferral algorithm in Telepoison itself so that it can be "ready-to-use" without all users having to code their own inferral functions (but I can be convinced otherwise). Obviously if we were to change the default inferral behaviour we would mark that as a breaking change.

**Comment by milquetoastable** [comment]
@LeartS @tsloughter I've implemented the above!

**Review by LeartS — COMMENTED** [review]
Nice work, a comment on the API, feel free to challenge it!

File: README.md [hunk 10]
```diff
@@ -10,33 +10,58 @@ Telepoison is a [opentelemetry-instrumented](https://github.com/open-telemetry/o
...
 
-Additionally, telepoison adds some options that can be passed in the `opts` HTTPoison argument to set OpenTelemetry-related stuff.
-These start with `:ot_`
+## Configuration
+
+`Telepoison.setup/1` takes a `Keyword list` that can configure if and how the `http.route` Open Telemetry metadata will be set per request using the `:infer_route` option.
+
+* If no value is provided then the out of the box, conservative inference provided by `Telepoison.URI.infer_route_from_request/1` is used to determine the inference
+
+* If an function with an arity of 1 (the `t:HTTPoison.Request/0` `request`) is provided then that function is used to determine the inference
```
Comment (cpiemontese) [inline README.md:31]:
```suggestion
* If a function with an arity of 1 (the `t:HTTPoison.Request/0` `request`) is provided then that function is used to determine the inference
```

Comment (cpiemontese) [reply README.md:31]: Also, if feel it could be clearer that `HTTPoison.Request` is the input of the unary function

File: README.md [hunk 11]
```diff
@@ -10,33 +10,58 @@ Telepoison is a [opentelemetry-instrumented](https://github.com/open-telemetry/o
...
+
+This can be overridden per each call to `Telepoison.request/1` derived functions (`Telepoison.get/3`, `Telepoison.get!/3` etc.)
+
```
Comment (cpiemontese) [inline README.md:34]: It would be cool to insert examples of this configurations, wdyt?

File: README.md [hunk 12]
```diff
@@ -10,33 +10,58 @@ Telepoison is a [opentelemetry-instrumented](https://github.com/open-telemetry/o
...
+Additionally, `Telepoison` provides some options that can be added to each derived function via
+the `Keyword list` `opts` parameter (or the `HTTPoison.Request` `options` `Keyword list` if calling `Telepoison.Request/1` directly). These are prefixed with `:ot_`.
 
+* `:ot_span_name` - sets the span name.
+* `:ot_attributes` - a list of `{name, value}` `tuple` attributes that will be added to the span.
+* `:ot_resource_route` - sets the `http.route` attribute, depending on the value provided.
+
+If the value is a string or an function with an arity of 1 (the `t:HTTPoison.Request/0` `request`) that is used to set the attribute
+
+If `:infer` is provided, then the function discussed within the *Configuration* section is used to set the attribute
```
Comment (cpiemontese) [inline README.md:46]:
```suggestion
If `:infer` is provided, then the function discussed within the (*Configuration*)[#configuration] section is used to set the attribute
```

I don't know if this is the exact syntax, but I'd add a link to the section if possible

**Review by cpiemontese — COMMENTED** [review]
Added a few minor comments, feel free to disregard them, otherwise looks good :smile: 

Just a general question: why is the default infer route limited to routes like `/segment` and `/segment1/segment2`? Maybe it's not a good API design, but what if I have `/user/:id/home` or whatever?

**Comment by tsloughter** [comment]
@LeartS you are probably right about per-instance.

**Comment by milquetoastable** [comment]
> Added a few minor comments, feel free to disregard them, otherwise looks good 😄
> 
> Just a general question: why is the default infer route limited to routes like `/segment` and `/segment1/segment2`? Maybe it's not a good API design, but what if I have `/user/:id/home` or whatever?

Thanks for the feedback! Yeah, we basically want the inference to be as cautious and conservative as possibly - callers _should_ explicitly provide the route in most cases, it's mainly to act as fallback so at least _something_ is registered

**Comment by cpiemontese** [comment]
@milquetoastable ok, I get it :+1: thanks for the explanation

File: README.md (hunk 12 above)
Comment (milquetoastable) [reply README.md]: Updated!

File: README.md (hunk 11 above)
Comment (milquetoastable) [reply README.md:36]: Added more examples a bit further down in the README.md

File: README.md (hunk 10 above)
Comment (milquetoastable) [reply README.md]: Updated, let me know if that's clearer!

**Review by cpiemontese — APPROVED** [review]
Looks good to me :+1:

File: README.md [hunk 13]
```diff
@@ -10,33 +10,103 @@ Telepoison is a [opentelemetry-instrumented](https://github.com/open-telemetry/o
...
+  "https://www.example.com/user/list",
   [],
-  ot_span_name: "fetch wikipedia homepage",
-  ot_attributes: [{"wikipedia.language", "en"}]
+  ot_resource_route: "my secret path"
 )
 ```
 
+In the example above:
+* `Telepoison.setup/1` is called with the `:infer_route` `Keyword list` option set to a function which takes a `%HTTPoison.Request/0` argument, returning the path of the request URL
```
Comment (LeartS) [inline README.md:105]: Isn't in this example `Telepoison.setup` called with no arguments?

File: README.md [hunk 14]
```diff
@@ -10,33 +10,103 @@ Telepoison is a [opentelemetry-instrumented](https://github.com/open-telemetry/o
...
+* `"my secret path"` is passed as the value for `:ot_resource_route` `Keyword list` option
+
+Given the above, the `http.route` attribute will be inferred as *my secret path*
```
Comment (LeartS) [inline README.md:108]: I wouldn't say in this example the attribute is _inferred_ as `my secret path`, but it is set explicitly

**Review by LeartS — APPROVED** [review]
Looks good, some minor nonblocking comments.

Also, are we missing an option to opt out of the `http.route` attribute entirely?

**Comment by milquetoastable** [comment]
> Looks good, some minor nonblocking comments.
> 
> Also, are we missing an option to opt out of the `http.route` attribute entirely?

Good point! Should this be configurable at a global level as well as per call?
